File: pyserver/slam_pipeline/management/commands/process_slam_jobs.py
"""
Port of bin/process_slam_jobs.php. Combines two patterns already proven
elsewhere in this project:

 - Atomic job claim via select_for_update(skip_locked=True) in a
   transaction (same trick as rinex's process_rinex_requests) -- guards
   against double-processing when scheduled runs overlap.
 - Spawning a step as a DETACHED subprocess with no wait (same trick as
   tours' process_copc_conversions/process_splat_transforms) -- SLAM steps
   can run for minutes/hours, the orchestrator must not block on them.

All DB logic lives in this command; the worker (slam_step_worker,
dispatching into slam_pipeline.pipeline.steps) knows nothing about the
database -- it only reads a context JSON and writes .done/.error files
(same sidecar principle as the .lock/.error files in the COPC/splat
pipelines), so Postgres access from the step implementations stays
centralized here.
"""

import json
import subprocess
import sys
import time
from pathlib import Path

from django.conf import settings
from django.core.management.base import BaseCommand
from django.db import transaction
from django.utils import timezone

from slam_pipeline.models import PIPELINE_ORDER, SlamJob, SlamProcessedAsset, SlamProject, SlamScan, SlamScanInput
from slam_pipeline.services import (
    slam_intermediate_path, slam_link_tour, slam_next_step,
    slam_scan_dir, slam_scan_intermediate_dir, slam_scan_jobs_dir, slam_scan_logs_dir,
)

# Same rationale as MAX_CONCURRENT=1 in the splat-transform pipeline (see
# tours/management/commands/process_splat_transforms.py) -- SLAM steps
# (especially compute_slam) are potentially just as memory-heavy, and the
# server has already gone down twice from parallel heavy conversions. No
# measured per-step memory profile yet -- not risking it.
MAX_CONCURRENT = 1
STALL_MINUTES = 120
STARTUP_GRACE_MINUTES = 15
MAX_STALL_RETRIES = 3


def _job_paths(scan_id: int, step: str) -> dict:
    jobs_dir = slam_scan_jobs_dir(scan_id)
    return {
        "dir": jobs_dir,
        "lock": jobs_dir / f"{step}.lock",
        "error": jobs_dir / f"{step}.error",
        "done": jobs_dir / f"{step}.done",
        "context": jobs_dir / f"{step}.context.json",
    }


class Command(BaseCommand):
    help = "Reconcile in-flight SLAM jobs and spawn new ones up to MAX_CONCURRENT."

    def handle(self, *args, **options):
        slam_upload_root = Path(settings.UPLOADS_ROOT) / "slam"
        slam_upload_root.mkdir(parents=True, exist_ok=True)

        self._reconcile_processing()
        spawned = self._spawn_new()
        self.stdout.write(f"Запущено новых job'ов: {spawned}")

    # -------------------------------------------------------------------
    # Pass 1: sort out "processing" jobs -- finished (done/error) or stalled
    # (lock disappeared/hasn't been touched in over STALL_MINUTES without a
    # result).
    # -------------------------------------------------------------------
    def _reconcile_processing(self):
        for job in SlamJob.objects.filter(status="processing").order_by("id"):
            scan_id = job.scan_id
            step = job.pipeline_step
            paths = _job_paths(scan_id, step)

            if paths["error"].is_file():
                msg = paths["error"].read_text(encoding="utf-8", errors="replace")[:2000]
                job.status, job.error_message, job.finished_at = "error", msg, timezone.now()
                job.save(update_fields=["status", "error_message", "finished_at"])
                SlamScan.objects.filter(id=scan_id).update(status="failed", error_message=msg)
                for p in ("error", "lock", "context"):
                    paths[p].unlink(missing_ok=True)
                self.stderr.write(f"Скан #{scan_id}, шаг {step}: ОШИБКА — {msg}")
                continue

            if paths["done"].is_file():
                self._finish_job(job, scan_id, step, paths)
                continue

            # Neither .done nor .error -- either still running, or stalled
            # (process killed by a crash/reboot before its finally ran).
            # Same rationale as in process_copc_conversions.py/
            # process_splat_transforms.py.
            lock_age_min = (time.time() - paths["lock"].stat().st_mtime) / 60 if paths["lock"].is_file() else None
            started_age_min = (timezone.now() - job.started_at).total_seconds() / 60 if job.started_at else 0
            stalled = False
            if lock_age_min is None and started_age_min > STARTUP_GRACE_MINUTES:
                stalled = True
                self.stderr.write(f"Скан #{scan_id}, шаг {step}: завис (лок пропал без результата спустя {round(started_age_min)} мин)")
            elif lock_age_min is not None and lock_age_min > STALL_MINUTES:
                stalled = True
                self.stderr.write(f"Скан #{scan_id}, шаг {step}: завис (лок не обновлялся дольше {STALL_MINUTES} мин)")

            if stalled:
                paths["lock"].unlink(missing_ok=True)
                paths["context"].unlink(missing_ok=True)
                if job.stall_retries + 1 >= MAX_STALL_RETRIES:
                    job.status = "error"
                    job.error_message = f"Шаг зависал {job.stall_retries} раз подряд (сервер перезапускался/процесс убит без завершения) — автоматические попытки остановлены."
                    job.finished_at = timezone.now()
                    job.save(update_fields=["status", "error_message", "finished_at"])
                    self.stderr.write(f"Скан #{scan_id}, шаг {step}: превышен лимит автоповторов ({job.stall_retries}), помечен как error")
                else:
                    job.status = "pending"
                    job.stall_retries += 1
                    job.started_at = None
                    job.save(update_fields=["status", "stall_retries", "started_at"])
                    self.stdout.write(f"Скан #{scan_id}, шаг {step}: будет перезапущен (попытка {job.stall_retries}/{MAX_STALL_RETRIES})")

    def _finish_job(self, job, scan_id, step, paths):
        meta = {}
        try:
            meta = json.loads(paths["done"].read_text(encoding="utf-8") or "{}")
        except (OSError, ValueError):
            pass

        with transaction.atomic():
            job.status, job.finished_at = "done", timezone.now()
            job.save(update_fields=["status", "finished_at"])

            scan_updates = {}
            if "num_points" in meta:
                scan_updates["num_points"] = int(meta["num_points"])
            if "source_format" in meta:
                scan_updates["source_format"] = str(meta["source_format"])
            if "crs_epsg" in meta:
                scan_updates["crs_epsg"] = int(meta["crs_epsg"])
            if "crs_proj4" in meta:
                scan_updates["crs_proj4"] = str(meta["crs_proj4"])
            if "rtk_fixed" in meta:
                scan_updates["rtk_fixed"] = 1 if meta["rtk_fixed"] else 0
            bbox = meta.get("bbox")
            if isinstance(bbox, list) and len(bbox) == 4:
                scan_updates["bbox_min_lon"], scan_updates["bbox_min_lat"], scan_updates["bbox_max_lon"], scan_updates["bbox_max_lat"] = bbox
            if scan_updates:
                SlamScan.objects.filter(id=scan_id).update(**scan_updates)

            for asset in meta.get("assets") or []:
                SlamProcessedAsset.objects.create(
                    scan_id=scan_id, asset_type=asset["type"], step=step,
                    file_path=asset["path"], file_size=int(asset.get("size") or 0),
                )

            for inp in meta.get("inputs") or []:
                SlamScanInput.objects.update_or_create(
                    scan_id=scan_id, kind=inp["kind"],
                    defaults={"file_path": inp["path"], "file_size": int(inp.get("size") or 0)},
                )

            next_step = slam_next_step(step)
            if next_step is not None:
                SlamJob.objects.get_or_create(scan_id=scan_id, pipeline_step=next_step, defaults={"status": "pending"})
            else:
                SlamScan.objects.filter(id=scan_id).update(status="completed")

        # build_octree is the last step: link the finished COPC to the
        # already-existing viewer (Фаза 5 -- the viewer already does
        # everything needed, "just wire it up"). tours.file_path is a
        # relative path inside uploads/tours/, not uploads/slam/, so we
        # copy (not move -- the original stays in slam_processed_assets)
        # the finished COPC into uploads/tours/slam_<scan_id>/ and create/
        # update one tours row per scan.
        if step == "build_octree":
            slam_link_tour(scan_id)

        for p in ("done", "lock", "context"):
            paths[p].unlink(missing_ok=True)
        next_step = slam_next_step(step)
        self.stdout.write(f"Скан #{scan_id}, шаг {step}: готово" + (f", следующий шаг: {next_step}" if next_step else " (пайплайн завершён)"))

    # -------------------------------------------------------------------
    # Pass 2: spawn new jobs (within MAX_CONCURRENT).
    # -------------------------------------------------------------------
    def _spawn_new(self) -> int:
        running_count = SlamJob.objects.filter(status="processing").count()
        self.stdout.write(f"Уже выполняется: {running_count} (лимит {MAX_CONCURRENT})")

        spawned = 0
        upload_dir = Path(settings.UPLOADS_ROOT) / "slam"
        while running_count + spawned < MAX_CONCURRENT:
            with transaction.atomic():
                claim = (
                    SlamJob.objects.select_for_update(skip_locked=True)
                    .filter(status="pending").order_by("id").first()
                )
                if not claim:
                    break
                job_id, scan_id, step = claim.id, claim.scan_id, claim.pipeline_step
                claim.status, claim.started_at, claim.updated_at, claim.error_message = "processing", timezone.now(), timezone.now(), None
                claim.save(update_fields=["status", "started_at", "updated_at", "error_message"])
                SlamScan.objects.filter(id=scan_id).update(status="processing")

            paths = _job_paths(scan_id, step)
            paths["dir"].mkdir(parents=True, exist_ok=True)
            slam_scan_logs_dir(scan_id).mkdir(parents=True, exist_ok=True)
            slam_scan_intermediate_dir(scan_id).mkdir(parents=True, exist_ok=True)

            scan = SlamScan.objects.get(id=scan_id)
            project = SlamProject.objects.get(id=scan.project_id)
            input_rows = list(SlamScanInput.objects.filter(scan_id=scan_id).values("kind", "file_path", "file_size"))

            context = {
                "scan_id": scan_id,
                "step": step,
                "scan_dir": str(slam_scan_dir(scan_id)),
                "raw_file_path": str(upload_dir / scan.raw_file_path) if scan.raw_file_path else None,
                "bag_lidar_enabled": bool(scan.bag_lidar_enabled),
                "skip_georeference": bool(scan.skip_georeference),
                "target_crs_epsg": project.target_crs_epsg,
                "target_crs_wkt": project.target_crs_wkt,
                "crs_proj4": scan.crs_proj4,
                "inputs": [{"kind": r["kind"], "path": str(upload_dir / r["file_path"])} for r in input_rows],
                "intermediate": {s: str(slam_intermediate_path(scan_id, s)) for s in PIPELINE_ORDER},
                "done_file": str(paths["done"]),
                "error_file": str(paths["error"]),
            }
            paths["context"].write_text(json.dumps(context, ensure_ascii=False), encoding="utf-8")
            paths["lock"].touch()

            log_file = slam_scan_logs_dir(scan_id) / f"{step}_{job_id}.log"
            with open(log_file, "a") as log_fh:
                subprocess.Popen(
                    [
                        sys.executable, str(Path(settings.BASE_DIR) / "manage.py"), "slam_step_worker",
                        "--step", step,
                        "--context-file", str(paths["context"]),
                        "--done-file", str(paths["done"]),
                        "--error-file", str(paths["error"]),
                        "--lock-file", str(paths["lock"]),
                    ],
                    stdin=subprocess.DEVNULL, stdout=log_fh, stderr=log_fh,
                    start_new_session=True,
                )
            spawned += 1
            self.stdout.write(f"Запущен шаг: скан #{scan_id}, {step} (job #{job_id})")

        return spawned

File: bin/process_slam_jobs.php
<?php
declare(strict_types=1);

/**
 * Оркестрация SLAM-пайплайна (перенос slamcloude, см. app/lib/slam.php для
 * порядка шагов). Сочетает два уже проверенных в проекте паттерна:
 *
 *  - Атомарный захват job'а SELECT...FOR UPDATE SKIP LOCKED в транзакции
 *    (тот же приём, что bin/process_rinex_requests.php) — защищает от
 *    двойной обработки при наложении запусков по расписанию.
 *  - Спавн шага как ОТДЕЛЬНОГО detached-процесса без ожидания (тот же
 *    приём, что bin/process_copc_conversions.php/process_splat_transforms.php)
 *    — шаги SLAM могут идти минуты/часы, оркестратор не должен блокироваться.
 *
 * DB-логика целиком в PHP; воркер (bin/slam_step_worker.ps1 -> node.exe
 * slam/steps/<step>.mjs) ничего не знает про БД — только читает контекстный
 * JSON и пишет .done/.error файлы (тот же sidecar-принцип, что .lock/.error
 * у COPC/сплат-пайплайнов), чтобы не плодить доступ к Postgres из Node.
 *
 * Запускать по расписанию (Windows Task Scheduler), как и остальные bin/*.php.
 */

require __DIR__ . '/../app/lib/db.php';
require __DIR__ . '/../app/lib/cli.php';
require __DIR__ . '/../app/lib/slam.php';
require_cli_or_token();

// Столько же обоснование, что MAX_CONCURRENT=1 у сплат-пайплайна (см.
// process_splat_transforms.php) — SLAM-шаги (особенно будущий compute_slam,
// Фаза 3) потенциально так же тяжелы по памяти, и сервер уже дважды падал
// от параллельных тяжёлых конвертаций. Пока нет измеренного профиля
// потребления памяти каждого шага — не рискуем.
const MAX_CONCURRENT = 1;
const STALL_MINUTES = 120;
const STARTUP_GRACE_MINUTES = 15;
const MAX_STALL_RETRIES = 3;

$psScript = __DIR__ . '/slam_step_worker.ps1';
$slamUploadRoot = __DIR__ . '/../uploads/slam';
if (!is_dir($slamUploadRoot)) {
    // На свежем сервере эта папка может ещё не существовать (ни один скан
    // не загружался через slam_scans.php) — без неё realpath() ниже вернёт
    // false и сломает построение путей контекста.
    mkdir($slamUploadRoot, 0777, true);
}
$pdo = db();

function job_paths(int $scanId, string $step): array
{
    $jobsDir = slam_scan_jobs_dir($scanId);
    return [
        'dir' => $jobsDir,
        'lock' => $jobsDir . '/' . $step . '.lock',
        'error' => $jobsDir . '/' . $step . '.error',
        'done' => $jobsDir . '/' . $step . '.done',
        'context' => $jobsDir . '/' . $step . '.context.json',
    ];
}

// ---------------------------------------------------------------------------
// Проход 1: разобрать уже "processing" job'ы — завершились (done/error) или
// зависли (лок пропал/не обновляется дольше STALL_MINUTES без результата).
// ---------------------------------------------------------------------------
$processing = $pdo->query(
    "SELECT * FROM slam_jobs WHERE status = 'processing' ORDER BY id"
)->fetchAll();

foreach ($processing as $job) {
    $scanId = (int)$job['scan_id'];
    $step = $job['pipeline_step'];
    $jobId = (int)$job['id'];
    $paths = job_paths($scanId, $step);

    if (is_file($paths['error'])) {
        $msg = substr((string)file_get_contents($paths['error']), 0, 2000);
        $pdo->prepare(
            "UPDATE slam_jobs SET status = 'error', error_message = :msg, finished_at = NOW() WHERE id = :id"
        )->execute(['msg' => $msg, 'id' => $jobId]);
        $pdo->prepare("UPDATE slam_scans SET status = 'failed', error_message = :msg WHERE id = :id")
            ->execute(['msg' => $msg, 'id' => $scanId]);
        @unlink($paths['error']);
        @unlink($paths['lock']);
        @unlink($paths['context']);
        cli_err("Скан #$scanId, шаг $step: ОШИБКА — $msg");
        continue;
    }

    if (is_file($paths['done'])) {
        $meta = json_decode((string)file_get_contents($paths['done']), true) ?: [];
        $pdo->beginTransaction();
        $pdo->prepare("UPDATE slam_jobs SET status = 'done', finished_at = NOW() WHERE id = :id")
            ->execute(['id' => $jobId]);

        $updates = [];
        $params = ['id' => $scanId];
        if (isset($meta['num_points'])) { $updates[] = 'num_points = :num_points'; $params['num_points'] = (int)$meta['num_points']; }
        if (isset($meta['source_format'])) { $updates[] = 'source_format = :source_format'; $params['source_format'] = (string)$meta['source_format']; }
        if (isset($meta['crs_epsg'])) { $updates[] = 'crs_epsg = :crs_epsg'; $params['crs_epsg'] = (int)$meta['crs_epsg']; }
        if (isset($meta['crs_proj4'])) { $updates[] = 'crs_proj4 = :crs_proj4'; $params['crs_proj4'] = (string)$meta['crs_proj4']; }
        if (isset($meta['rtk_fixed'])) { $updates[] = 'rtk_fixed = :rtk_fixed'; $params['rtk_fixed'] = $meta['rtk_fixed'] ? 1 : 0; }
        if (isset($meta['bbox']) && is_array($meta['bbox']) && count($meta['bbox']) === 4) {
            $updates[] = 'bbox_min_lon = :bminlon, bbox_min_lat = :bminlat, bbox_max_lon = :bmaxlon, bbox_max_lat = :bmaxlat';
            [$params['bminlon'], $params['bminlat'], $params['bmaxlon'], $params['bmaxlat']] = $meta['bbox'];
        }
        if ($updates) {
            $pdo->prepare('UPDATE slam_scans SET ' . implode(', ', $updates) . ' WHERE id = :id')->execute($params);
        }

        foreach ((array)($meta['assets'] ?? []) as $asset) {
            $pdo->prepare(
                'INSERT INTO slam_processed_assets (scan_id, asset_type, step, file_path, file_size)
                 VALUES (:scan_id, :asset_type, :step, :file_path, :file_size)'
            )->execute([
                'scan_id' => $scanId,
                'asset_type' => $asset['type'],
                'step' => $step,
                'file_path' => $asset['path'],
                'file_size' => (int)($asset['size'] ?? 0),
            ]);
        }

        foreach ((array)($meta['inputs'] ?? []) as $input) {
            $pdo->prepare(
                'INSERT INTO slam_scan_inputs (scan_id, kind, file_path, file_size)
                 VALUES (:scan_id, :kind, :file_path, :file_size)
                 ON CONFLICT (scan_id, kind) DO UPDATE SET file_path = EXCLUDED.file_path, file_size = EXCLUDED.file_size'
            )->execute([
                'scan_id' => $scanId,
                'kind' => $input['kind'],
                'file_path' => $input['path'],
                'file_size' => (int)($input['size'] ?? 0),
            ]);
        }

        $next = slam_next_step($step);
        if ($next !== null) {
            $pdo->prepare(
                "INSERT INTO slam_jobs (scan_id, pipeline_step, status) VALUES (:scan_id, :step, 'pending')
                 ON CONFLICT (scan_id, pipeline_step) DO NOTHING"
            )->execute(['scan_id' => $scanId, 'step' => $next]);
        } else {
            $pdo->prepare("UPDATE slam_scans SET status = 'completed' WHERE id = :id")->execute(['id' => $scanId]);
        }
        $pdo->commit();

        // build_octree — последний шаг: связываем готовый COPC с уже
        // существующим вьювером (Фаза 5 плана — вьювер уже умеет всё
        // нужное, "просто связать"). tours.file_path — относительный путь
        // внутри uploads/tours/, а не uploads/slam/, поэтому копируем
        // (не перемещаем — оригинал остаётся в slam_processed_assets)
        // готовый COPC в uploads/tours/slam_<scan_id>/ и создаём/обновляем
        // одну запись tours на скан.
        if ($step === 'build_octree') {
            slam_link_tour($pdo, $scanId);
        }

        @unlink($paths['done']);
        @unlink($paths['lock']);
        @unlink($paths['context']);
        cli_out("Скан #$scanId, шаг $step: готово" . ($next ? ", следующий шаг: $next" : ' (пайплайн завершён)'));
        continue;
    }

    // Ни .done, ни .error — либо ещё работает, либо зависла (процесс убит
    // сбоем/перезагрузкой до finally). См. то же обоснование в
    // process_copc_conversions.php/process_splat_transforms.php.
    $lockAgeMin = is_file($paths['lock']) ? (time() - filemtime($paths['lock'])) / 60 : null;
    $startedAgeMin = $job['started_at'] ? (time() - strtotime($job['started_at'])) / 60 : 0;
    $stalled = false;
    if ($lockAgeMin === null && $startedAgeMin > STARTUP_GRACE_MINUTES) {
        $stalled = true;
        cli_err("Скан #$scanId, шаг $step: завис (лок пропал без результата спустя " . round($startedAgeMin) . " мин)");
    } elseif ($lockAgeMin !== null && $lockAgeMin > STALL_MINUTES) {
        $stalled = true;
        cli_err("Скан #$scanId, шаг $step: завис (лок не обновлялся дольше " . STALL_MINUTES . " мин)");
    }

    if ($stalled) {
        $retries = (int)$job['stall_retries'];
        @unlink($paths['lock']);
        @unlink($paths['context']);
        if ($retries + 1 >= MAX_STALL_RETRIES) {
            $pdo->prepare(
                "UPDATE slam_jobs SET status = 'error', error_message = :msg, finished_at = NOW() WHERE id = :id"
            )->execute([
                'msg' => "Шаг зависал $retries раз подряд (сервер перезапускался/процесс убит без завершения) — автоматические попытки остановлены.",
                'id' => $jobId,
            ]);
            cli_err("Скан #$scanId, шаг $step: превышен лимит автоповторов ($retries), помечен как error");
        } else {
            $pdo->prepare(
                "UPDATE slam_jobs SET status = 'pending', stall_retries = :retries, started_at = NULL WHERE id = :id"
            )->execute(['retries' => $retries + 1, 'id' => $jobId]);
            cli_out("Скан #$scanId, шаг $step: будет перезапущен (попытка " . ($retries + 1) . '/' . MAX_STALL_RETRIES . ')');
        }
    }
}

// ---------------------------------------------------------------------------
// Проход 2: спавн новых job'ов (в пределах MAX_CONCURRENT).
// ---------------------------------------------------------------------------
$runningCount = (int)$pdo->query("SELECT COUNT(*) FROM slam_jobs WHERE status = 'processing'")->fetchColumn();
cli_out("Уже выполняется: $runningCount (лимит " . MAX_CONCURRENT . ')');

$spawned = 0;
while ($runningCount + $spawned < MAX_CONCURRENT) {
    $pdo->beginTransaction();
    $claim = $pdo->query(
        "SELECT * FROM slam_jobs WHERE status = 'pending' ORDER BY id FOR UPDATE SKIP LOCKED LIMIT 1"
    )->fetch();
    if (!$claim) {
        $pdo->rollBack();
        break;
    }
    $jobId = (int)$claim['id'];
    $scanId = (int)$claim['scan_id'];
    $step = $claim['pipeline_step'];
    $pdo->prepare(
        "UPDATE slam_jobs SET status = 'processing', started_at = NOW(), updated_at = NOW(), error_message = NULL WHERE id = :id"
    )->execute(['id' => $jobId]);
    $pdo->prepare("UPDATE slam_scans SET status = 'processing' WHERE id = :id")->execute(['id' => $scanId]);
    $pdo->commit();

    $paths = job_paths($scanId, $step);
    if (!is_dir($paths['dir'])) {
        mkdir($paths['dir'], 0777, true);
    }
    $logsDir = slam_scan_logs_dir($scanId);
    if (!is_dir($logsDir)) {
        mkdir($logsDir, 0777, true);
    }
    if (!is_dir(slam_scan_intermediate_dir($scanId))) {
        mkdir(slam_scan_intermediate_dir($scanId), 0777, true);
    }

    // Контекст для воркера: всё, что шаг может захотеть прочитать — сам
    // PHP единственный, кто трогает БД, воркер работает только с файлами.
    $scan = $pdo->prepare('SELECT * FROM slam_scans WHERE id = :id');
    $scan->execute(['id' => $scanId]);
    $scanRow = $scan->fetch();
    $project = $pdo->prepare('SELECT * FROM slam_projects WHERE id = :id');
    $project->execute(['id' => (int)$scanRow['project_id']]);
    $projectRow = $project->fetch();
    $inputs = $pdo->prepare('SELECT kind, file_path, file_size FROM slam_scan_inputs WHERE scan_id = :id');
    $inputs->execute(['id' => $scanId]);
    $inputRows = $inputs->fetchAll();

    $uploadDir = realpath(__DIR__ . '/../uploads/slam') . DIRECTORY_SEPARATOR;
    $context = [
        'scan_id' => $scanId,
        'step' => $step,
        'scan_dir' => slam_scan_dir($scanId),
        'raw_file_path' => $scanRow['raw_file_path'] ? ($uploadDir . $scanRow['raw_file_path']) : null,
        'bag_lidar_enabled' => (bool)$scanRow['bag_lidar_enabled'],
        'skip_georeference' => (bool)$scanRow['skip_georeference'],
        'target_crs_epsg' => $projectRow['target_crs_epsg'] !== null ? (int)$projectRow['target_crs_epsg'] : null,
        'target_crs_wkt' => $projectRow['target_crs_wkt'],
        'crs_proj4' => $scanRow['crs_proj4'],
        'inputs' => array_map(fn($r) => [
            'kind' => $r['kind'],
            'path' => $uploadDir . $r['file_path'],
        ], $inputRows),
        'intermediate' => array_reduce(SLAM_PIPELINE_ORDER, function ($acc, $s) use ($scanId) {
            $acc[$s] = slam_intermediate_path($scanId, $s);
            return $acc;
        }, []),
        'done_file' => $paths['done'],
        'error_file' => $paths['error'],
    ];
    file_put_contents($paths['context'], json_encode($context, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES));
    touch($paths['lock']);

    $logFile = $logsDir . '/' . $step . '_' . $jobId . '.log';
    $cmd = [
        'powershell.exe', '-NoProfile', '-ExecutionPolicy', 'Bypass', '-File', $psScript,
        '-Step', $step,
        '-ContextFile', $paths['context'],
        '-DoneFile', $paths['done'],
        '-ErrorFile', $paths['error'],
        '-LockFile', $paths['lock'],
    ];
    $descriptors = [0 => ['pipe', 'r'], 1 => ['file', $logFile, 'a'], 2 => ['file', $logFile, 'a']];
    $process = proc_open($cmd, $descriptors, $pipes);
    if ($process === false) {
        cli_err("Не удалось запустить шаг $step для скана #$scanId");
        $pdo->prepare("UPDATE slam_jobs SET status = 'pending', started_at = NULL WHERE id = :id")->execute(['id' => $jobId]);
        @unlink($paths['lock']);
        break;
    }
    fclose($pipes[0]);
    $spawned++;
    cli_out("Запущен шаг: скан #$scanId, $step (job #$jobId)");
}

cli_out("Запущено новых job'ов: $spawned");
